File: clinical-schema/js/src/builders/observation.js
import { VALTYPE, DEFAULT_SOURCESYSTEM_CD, DEFAULT_PROVIDER_ID } from '../constants.js'
import { isoNow, isoDate, stringifyBlob } from './util.js'

/**
 * Build an OBSERVATION_FACT record.
 *
 * Pass `value` for convenience: numbers populate NVAL_NUM with VALTYPE_CD='N',
 * strings populate TVAL_CHAR with VALTYPE_CD='T'. Override VALTYPE_CD/NVAL_NUM/
 * TVAL_CHAR explicitly when needed.
 *
 * Required: OBSERVATION_ID, ENCOUNTER_NUM, PATIENT_NUM, CONCEPT_CD.
 */
export function buildObservation({
  OBSERVATION_ID,
  ENCOUNTER_NUM,
  PATIENT_NUM,
  CONCEPT_CD,
  CATEGORY_CHAR = null,
  PROVIDER_ID = DEFAULT_PROVIDER_ID,
  START_DATE = isoNow(),
  END_DATE = null,
  INSTANCE_NUM = 1,
  value = undefined,
  VALTYPE_CD = undefined,
  TVAL_CHAR = undefined,
  NVAL_NUM = undefined,
  UNIT_CD = null,
  QUANTITY_NUM = null,
  VALUEFLAG_CD = null,
  LOCATION_CD = null,
  CONFIDENCE_NUM = null,
  OBSERVATION_BLOB = null,
  SOURCESYSTEM_CD = DEFAULT_SOURCESYSTEM_CD,
  UPLOAD_ID = 1,
} = {}) {
  if (OBSERVATION_ID == null) throw new Error('buildObservation: OBSERVATION_ID is required')
  if (ENCOUNTER_NUM == null) throw new Error('buildObservation: ENCOUNTER_NUM is required')
  if (PATIENT_NUM == null) throw new Error('buildObservation: PATIENT_NUM is required')
  if (!CONCEPT_CD) throw new Error('buildObservation: CONCEPT_CD is required')

  if (VALTYPE_CD === undefined) {
    if (typeof value === 'number') {
      VALTYPE_CD = VALTYPE.NUMERIC
      NVAL_NUM = NVAL_NUM ?? value
      TVAL_CHAR = TVAL_CHAR ?? null
    } else if (typeof value === 'string') {
      VALTYPE_CD = VALTYPE.TEXT
      TVAL_CHAR = TVAL_CHAR ?? value
      NVAL_NUM = NVAL_NUM ?? null
    } else {
      VALTYPE_CD = VALTYPE.TEXT
    }
  }

  return {
    OBSERVATION_ID,
    ENCOUNTER_NUM,
    PATIENT_NUM,
    CATEGORY_CHAR,
    CONCEPT_CD,
    PROVIDER_ID,
    START_DATE,
    INSTANCE_NUM,
    VALTYPE_CD,
    TVAL_CHAR: TVAL_CHAR ?? null,
    NVAL_NUM: NVAL_NUM ?? null,
    VALUEFLAG_CD,
    QUANTITY_NUM,
    UNIT_CD,
    END_DATE,
    LOCATION_CD,
    CONFIDENCE_NUM,
    OBSERVATION_BLOB: stringifyBlob(OBSERVATION_BLOB),
    UPDATE_DATE: null,
    DOWNLOAD_DATE: null,
    IMPORT_DATE: null,
    SOURCESYSTEM_CD,
    UPLOAD_ID,
    CREATED_AT: isoDate(),
  }
}

File: clinical-schema/python/src/dbbest_clinical_schema/models/observation.py
"""Observation model — mirrors OBSERVATION_FACT."""
from __future__ import annotations

import json
from typing import Any, Optional, Union

from pydantic import Field, field_validator

from ._base import ClinicalBase, iso_now, iso_date
from ..constants import (
    DEFAULT_PROVIDER_ID,
    DEFAULT_SOURCESYSTEM_CD,
    VALTYPE_CODES,
    ValType,
)


class QuestionnaireBlob(ClinicalBase):
    """Shape of OBSERVATION_BLOB when VALTYPE_CD=='Q'."""

    label: Optional[str] = None
    title: Optional[str] = None
    short_title: Optional[str] = None
    questionnaire_code: Optional[str] = None
    date_start: Optional[int] = None
    date_end: Optional[int] = None
    items: list[dict[str, Any]] = Field(default_factory=list)
    results: list[dict[str, Any]] = Field(default_factory=list)
    coding: Optional[dict[str, Any]] = None


class Observation(ClinicalBase):
    OBSERVATION_ID: int
    ENCOUNTER_NUM: int
    PATIENT_NUM: int
    CONCEPT_CD: str
    CATEGORY_CHAR: Optional[str] = None
    PROVIDER_ID: str = DEFAULT_PROVIDER_ID
    START_DATE: str = Field(default_factory=iso_now)
    END_DATE: Optional[str] = None
    INSTANCE_NUM: int = 1
    VALTYPE_CD: str
    TVAL_CHAR: Optional[str] = None
    NVAL_NUM: Optional[float] = None
    VALUEFLAG_CD: Optional[str] = None
    QUANTITY_NUM: Optional[float] = None
    UNIT_CD: Optional[str] = None
    LOCATION_CD: Optional[str] = None
    CONFIDENCE_NUM: Optional[float] = None
    OBSERVATION_BLOB: Optional[str] = None
    UPDATE_DATE: Optional[str] = None
    DOWNLOAD_DATE: Optional[str] = None
    IMPORT_DATE: Optional[str] = None
    SOURCESYSTEM_CD: str = DEFAULT_SOURCESYSTEM_CD
    UPLOAD_ID: int = 1
    CREATED_AT: str = Field(default_factory=iso_date)

    @field_validator("VALTYPE_CD")
    @classmethod
    def _check_valtype(cls, v: str) -> str:
        if v not in VALTYPE_CODES:
            raise ValueError(f"VALTYPE_CD must be one of {VALTYPE_CODES}, got {v!r}")
        return v

    @field_validator("OBSERVATION_BLOB", mode="before")
    @classmethod
    def _stringify_blob(cls, v: Union[str, dict, list, None]) -> Optional[str]:
        if v is None:
            return None
        if isinstance(v, (dict, list)):
            return json.dumps(v)
        if isinstance(v, QuestionnaireBlob):
            return v.model_dump_json()
        return v

    def parsed_blob(self) -> Optional[dict[str, Any]]:
        """Parse OBSERVATION_BLOB as JSON, returning None if empty/invalid."""
        if not self.OBSERVATION_BLOB:
            return None
        try:
            return json.loads(self.OBSERVATION_BLOB)
        except (TypeError, ValueError):
            return None

    @property
    def is_questionnaire(self) -> bool:
        return self.VALTYPE_CD == ValType.QUESTIONNAIRE.value
